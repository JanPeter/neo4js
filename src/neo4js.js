import { v1 as neo4j } from 'neo4j-driver';
import Utils from './Utils';
import uuid from 'node-uuid';
import Debug from 'debug';
const debug = Debug('neo4js');

import * as neo4jsErrors from './errors';
import Model from './model';
import ModelManager from './model-manager';
import ModelObject from './model-object';
import Query from './query';
import { getSchemaInfo } from './schema-info';

class Neo4js {
  /**
   * Constructor to create a new instance of the Neo4js class
   * @param {String}  username The username which is used to authenticate against the database
   * @param {String}  password The password which is used to authenticate against the database
   * @param {Object}  [options={}] An object with options
   * @param {String}  [options.url=localhost] Bolt url
   * @param {Integer} [options.port=7687] Bolt port
   * @param {Object}  [options.rest={}]
   * @param {String}  [options.rest.url=localhost]
   * @param {Integer} [options.rest.port=7474]
   * @param {Integer} [options.connectionPoolSize=50]
   */
  constructor(username, password, options) {
    this.auth = { username, password };

    const defaultOptions = {
      url: `bolt://localhost`,
      port: 7687,
      rest: {
        url: `http://localhost`,
        port: 7474,
      },
      driver: {
        connectionPoolSize: 50,
      },
    };

    options = Utils._.extend({}, defaultOptions, options || {});
    options.rest = Utils._.extend({}, defaultOptions.rest, options.rest);
    options.driver = Utils._.extend({}, defaultOptions.driver, options.driver);

    // TODO: Maybe add here some more checks...
    const url = `${options.url}:${options.port}`;
    this.driver = neo4j.driver(url, neo4j.auth.basic(username, password), options.driver);

    this.options = options;

    this.models = {};
    this.modelManager = new ModelManager(this);
  }

  /**
   * @param {String | String[]} labels
   * @param {Object} schema
   * @param {Object} schema.{propertyName}
   * @param {Boolean} schema.{propertyName}.index
   * @param {Boolean} schema.{propertyName}.unique
   * @param {Boolean} schema.{propertyName}.exists
   * @param {Function} schema.{propertyName}.defaultValue
   */
  define(labels, schema, hooks) {
    const model = new Model(labels, schema, hooks, this);
    this.modelManager.add(model);
    return model;
  }

  /**
   * @param {String | String[]} labels
   */
  getModel(labels) {
    const model = this.modelManager.getModel(labels);
    return model;
  }

  /**
   * @returns {Promise}
   */
  getSchemaInfo() {
    return getSchemaInfo(this.options.rest.url, this.options.rest.port, this.auth);
  }

  drop() {
    return this.getSchemaInfo()
      .then(result => {
        const cmds = [];
        for (const label in result) {
          for (const p in result[label]) {
            if (result[label][p].index) {
              cmds.push(new Query(this).dropIndex(label, p));
            }
            if (result[label][p].unique) {
              cmds.push(new Query(this).dropConstraint(label, p, 'unique'));
            }
          }
        }
        cmds.push(new Query(this).match('n').detach('n'));
        return Query.runSequence(cmds);
      });
  }

  /**
   * @param {Object} options
   * @param {Boolean} options.force
   */
  sync(options) {
    debug(`sync start`);
    options = options || {};

    return Promise.resolve()
      .then(() => {
        if (options.force) {
          debug(`cleanup database start`);
          return this.drop();
        }
      })
      .then(() => {
        if (options.force) {
          debug(`cleanup database end`);
        }
        const promises = this.modelManager.all.map(m => m.sync());
        return Promise.all(promises);
      })
      .then(syncResults => {
        const errors = [];
        for (const results of syncResults) {
          for (const result of results) {
            if (result.err) {
              errors.push(result.err);
            }
          }
        }
        return errors;
      })
      .then(result => {
        debug(`sync end`);
        return result;
      })
      .catch(err => {
        debug(`sync end with errors`, err);
        return err;
      });
  }

  /**
   * This method creates a new instance of a Query
   * @return {Query}
   */
  Query() {
    return new Query(this);
  }

  /**
   * This method runs a command with params
   * @param {String} cmd - Cypher statement to execute
   * @param {Object} params - Map with parameters to use in statement
   * @return {Promise} Resolves either with the result or the error message
   */
  run(cmd, params) {
    const session = this.driver.session();
    return session.run(cmd, params)
      .then(result => {
        session.close();
        return result;
      })
      .catch(err => {
        session.close();
        return err;
      });
  }

  /**
   * @returns {UUIDv4}
   */
  static uuid4() {
    return uuid.v4();
  }

  /**
   * @returns {UUIDv1}
   */
  static uuid1() {
    return uuid.v1();
  }
}

Neo4js.prototype.Error = Neo4js.Error = neo4jsErrors.BaseError;
Neo4js.prototype.Model = Neo4js.Model = Model;
Neo4js.prototype.ModelObject = Neo4js.ModelObject = ModelObject;

export default Neo4js;
