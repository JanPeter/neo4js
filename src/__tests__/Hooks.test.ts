// @flow

import neo4js, { Model, ModelInstance } from "../index";

type Props = {
  name?: string;
  age?: number;
};

class PersonInstance extends ModelInstance<Props> {}
class PersonModel extends Model<Props, PersonInstance> {}
let Person = new PersonModel("Person");

type TaskProps = {
  title: string;
  done?: boolean;
};

class TaskInstance extends ModelInstance<TaskProps> {}
class TaskModel extends Model<TaskProps, TaskInstance> {
  afterFind(t: TaskInstance): TaskInstance {
    delete t.props.guid;
    t.props.done = t.props.done || false;
    return t;
  }
}
let Task = new TaskModel("Task");

describe("Hooks", () => {
  beforeAll(() => {
    neo4js.init({
      boltUri: "localhost",
      boltPort: 10001,
    });
  });

  beforeEach(() => {
    Person = new PersonModel("Person");
    Task = new TaskModel("Task");
  });

  afterEach(async () => {
    await neo4js.run("MATCH (n) DETACH DELETE n");
  });

  afterAll(() => {
    neo4js.close();
  });

  describe("create", () => {
    it("should call before create hook once", async () => {
      // @ts-ignore
      Person.beforeCreate = jest.fn().mockImplementationOnce(t => t);
      const paul = await Person.create({ name: "Paul", age: 21 });
      // @ts-ignore
      expect(Person.beforeCreate).toHaveBeenCalledTimes(1);
    });

    it("should call before create hook once and change props", async () => {
      // @ts-ignore
      Person.beforeCreate = (t: Props): Props => ({ ...t, age: t.age + 1 });
      const paul = await Person.create({ name: "Paul", age: 21 });
      expect(paul.props.age).toEqual(22);
    });

    it("should call before create hook once", async () => {
      // @ts-ignore
      Person.afterCreate = jest.fn().mockImplementationOnce(t => t);
      const paul = await Person.create({ name: "Paul", age: 21 });
      // @ts-ignore
      expect(Person.afterCreate).toHaveBeenCalledTimes(1);
    });

    it("should call after create hook once and change props of instance", async () => {
      // @ts-ignore
      Person.afterCreate = (t: PersonInstance) => {
        delete t.props.guid;
        return t;
      };
      const paul = await Person.create({ name: "Paul", age: 21 });
      expect(paul).toMatchSnapshot();
    });
  });

  describe("find", () => {
    it("should call before find hook once", async () => {
      // @ts-ignore
      Person.beforeFind = jest.fn().mockImplementationOnce(t => t);
      await Person.find();
      // @ts-ignore
      expect(Person.beforeFind).toHaveBeenCalledTimes(1);
    });

    it("should call before find hook once and change props", async () => {
      // @ts-ignore
      Task.beforeFind = (p: TaskProps) => ({ ...p, done: true });
      await Task.create({ title: "Buy milk" });
      await Task.create({ title: "Read a book", done: true });
      const tasks = await Task.find();
      expect(tasks).toMatchSnapshot();
    });

    it("should call after find hook twice", async () => {
      // @ts-ignore
      Person.afterFind = jest.fn().mockImplementation(t => t);
      await Person.create({ name: "Hanns" });
      await Person.create({ name: "Olaf" });
      await Person.find();
      // @ts-ignore
      expect(Person.afterFind).toHaveBeenCalledTimes(2);
    });

    it("should call after find hook twice and change props of instances", async () => {
      await Task.create({ title: "Buy milk" });
      await Task.create({ title: "Read a book" });
      const tasks = await Task.find();
      expect(tasks).toMatchSnapshot();
    });
  });

  describe("update", () => {
    it("should call before update hook once", async () => {
      // @ts-ignore
      Person.beforeUpdate = jest
        .fn()
        .mockImplementationOnce((props, newProps) => ({ props, newProps }));
      await Person.create({ name: "Hanns", age: 20 });
      await Person.update({ name: "Hanns" }, { age: 21 });
      // @ts-ignore
      expect(Person.beforeUpdate).toHaveBeenCalledTimes(1);
    });

    it("should call before find hook once and change props", async () => {
      // @ts-ignore
      Person.beforeUpdate = (props: Props, newProps: Props) => {
        newProps.age += 1;
        return { props, newProps };
      };
      await Person.create({ name: "Hanns", age: 20 });
      const p = (await Person.update({ name: "Hanns" }, { age: 21 }))[0];
      expect(p.props.age).toEqual(22);
    });

    it("should call after find hook twice", async () => {
      // @ts-ignore
      Person.afterUpdate = jest.fn().mockImplementation(t => t);
      await Person.create({ name: "Hanns" });
      await Person.create({ name: "Olaf" });
      await Person.update({ name: "Hanns" }, { age: 20 });
      // @ts-ignore
      expect(Person.afterUpdate).toHaveBeenCalledTimes(1);
    });

    // it('should call after find hook twice and change props of instances', async () => {
    //   await Task.create({ title: 'Buy milk' });
    //   await Task.create({ title: 'Read a book' });
    //   const tasks = await Task.find();
    //   expect(tasks).toMatchSnapshot();
    // });
  });
});
