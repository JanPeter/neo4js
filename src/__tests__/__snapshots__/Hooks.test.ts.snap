// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hooks create should call after create hook once and change props of instance 1`] = `
ModelInstance {
  "props": Object {
    "age": 21,
    "name": "Paul",
  },
}
`;

exports[`Hooks find should call after find hook twice and change props of instances 1`] = `
Array [
  ModelInstance {
    "props": Object {
      "done": false,
      "title": "Buy milk",
    },
  },
  ModelInstance {
    "props": Object {
      "done": false,
      "title": "Read a book",
    },
  },
]
`;

exports[`Hooks find should call before find hook once and change props 1`] = `
Array [
  ModelInstance {
    "props": Object {
      "done": true,
      "title": "Read a book",
    },
  },
]
`;
